### PR TITLE
Support for coroutines.

### LINKED ISSUE #150: Support for coroutines.
If one writes an `async def` method, which returns a coroutine, Klein doesn't handle that, so you have to figure out how to turn that into a non-async method that returns a `Deferred`.

For example, I do that with a decorator [here](https://github.com/wsanchez/sample-klein-app/blob/979b02c94fcd8d6caa5a0c40fe5ee575c0ca63a3/src/sample_klein_app/application/dns.py#L20).

It would be nicer if Klein just handled coroutines.

### PR DESCRIPTION
**Pull request by wsanchez**
Fixes #150

### CODE REVIEW

**Comment by wsanchez** [comment]
OK, this is working except for Py3.5 and Twisted 15.5.

The code is, I think, working fine; it's raising `NotImplementedError` as expected, but [my attempt to catch that error](https://github.com/twisted/klein/blob/a0e1e7003ae6e87769862e494e81e92433ca4ebd/src/klein/test/py3_test_resource.py#L35) in the test is failing to do so, and I can't tell why that is.

File: src/klein/test/test_resource.py [hunk 1]
```diff
@@ -1153,3 +1155,12 @@ def test_afUnixSocket(self):
         self.assertIsInstance(server_port, int)
         self.assertIsInstance(path_info, unicode)
         self.assertIsInstance(script_name, unicode)
+
+
+if _PY3:
+    from sys import version_info as pyVersion
+
+    if pyVersion.major > 3 or (pyVersion.major == 3 and pyVersion.minor >= 5):
```
Comment (notoriousno) [inline src/klein/test/test_resource.py]: For the most part, Py2 and Py3 syntax shouldn't be too different, so do we really need a separate file for Py3? If it's just housing the ``async`` syntax stuff then this is fine, but with an all-encompassing name like ``py3_test_resource.py``, I'm inclined to think all Py3 tests would go in there.

File: src/klein/test/test_resource.py [hunk 2]
```diff
@@ -1153,3 +1155,12 @@ def test_afUnixSocket(self):
...
+        from os.path import join, dirname
+        name = join(dirname(__file__), "py3_test_resource.py")
```
Comment (notoriousno) [inline src/klein/test/test_resource.py]: The name of the file implies all Py3 tests should go in there. This wouldn't be ideal considering Klein/Twisted is a single code base for Py2 and Py3. Maybe rename the file ``coroutine_tests.py`` or something along those lines.

File: src/klein/test/test_resource.py [hunk 3]
```diff
@@ -1153,3 +1155,12 @@ def test_afUnixSocket(self):
...
+        exec(open(name).read())
```
Comment (notoriousno) [inline src/klein/test/test_resource.py]: Yikes! Potential security risk.

**Comment by notoriousno** [comment]
I got a chance to test this out today on Py3.5. I think I may have an alternative solution for running awaitable tests. Instead of ``exec(open(name).read())``, let's simply import the test class:

``` python
if pyVersion.major > 3 or (pyVersion.major == 3 and pyVersion.minor >= 5):
    from klein.test.py3_test_resource import PY3KleinResourceTests
```

Sweet and simple

**Comment by wsanchez** [comment]
@notoriousno: I thought I had tried the simple import and it didn't work but I must have done that early when there were other issues in play.  I made the change you suggested and that worked.

I also cleaned up `test_asyncResourceRendering` which was failing on Py3.5 with Twisted 16.5 because catching rendering exceptions isn't as obvious as I thought.

If you could give this another look, that'd be rad.  Thanks.

**Pull request merged**